File: docs/07-Graphs-and-Disjoint-Set/python/dsu.py
class DSU:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, i):
        if self.parent[i] == i:
            return i
        # Path compression
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    def union(self, i, j):
        root_i = self.find(i)
        root_j = self.find(j)
        
        if root_i != root_j:
            # Union by rank
            if self.rank[root_i] < self.rank[root_j]:
                self.parent[root_i] = root_j
            elif self.rank[root_i] > self.rank[root_j]:
                self.parent[root_j] = root_i
            else:
                self.parent[root_j] = root_i
                self.rank[root_i] += 1
            return True
        return False

if __name__ == "__main__":
    dsu = DSU(5)
    print("Union(0, 2):", dsu.union(0, 2))
    print("Union(4, 2):", dsu.union(4, 2))
    print("Union(3, 1):", dsu.union(3, 1))
    
    print("Find(4) == Find(0):", dsu.find(4) == dsu.find(0)) # True
    print("Find(1) == Find(0):", dsu.find(1) == dsu.find(0)) # False

File: docs/07-Graphs-and-Disjoint-Set/js/dsu.js
class DSU {
    constructor(n) {
        this.parent = Array.from({ length: n }, (_, i) => i);
        this.rank = Array(n).fill(0);
    }

    find(i) {
        if (this.parent[i] === i) {
            return i;
        }
        return this.parent[i] = this.find(this.parent[i]); // Path compression
    }

    union(i, j) {
        const rootI = this.find(i);
        const rootJ = this.find(j);

        if (rootI !== rootJ) {
            if (this.rank[rootI] < this.rank[rootJ]) {
                this.parent[rootI] = rootJ;
            } else if (this.rank[rootI] > this.rank[rootJ]) {
                this.parent[rootJ] = rootI;
            } else {
                this.parent[rootJ] = rootI;
                this.rank[rootI]++;
            }
            return true;
        }
        return false;
    }
}

// Driver testing
const dsu = new DSU(5);
console.log("Union(0, 2):", dsu.union(0, 2));
console.log("Union(4, 2):", dsu.union(4, 2));
console.log("Union(3, 1):", dsu.union(3, 1));

console.log("Find(4) == Find(0):", dsu.find(4) === dsu.find(0)); // True
console.log("Find(1) == Find(0):", dsu.find(1) === dsu.find(0)); // False
